src/servers/src/row_writer.rs

```diff
@@ -0,0 +1,281 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use api::v1::value::ValueData;
+use api::v1::{
+    ColumnDataType, ColumnSchema, Row, RowInsertRequest, RowInsertRequests, Rows, SemanticType,
+    Value,
+};
+use common_grpc::writer;
+use common_grpc::writer::Precision;
+use common_time::timestamp::TimeUnit;
+use common_time::Timestamp;
+use snafu::{ensure, OptionExt, ResultExt};
+
+use crate::error::{IncompatibleSchemaSnafu, InfluxdbLinesWriteSnafu, Result, TimePrecisionSnafu};
+
+pub struct TableData<'a> {
+    schema: Vec<ColumnSchema>,
+    rows: Vec<Row>,
+    column_indexes: HashMap<&'a str, usize>,
+}
+
+impl TableData<'_> {
+    pub fn new(num_columns: usize, num_rows: usize) -> Self {
+        Self {
+            schema: if num_columns > 0 {
+                Vec::with_capacity(num_columns)
+            } else {
+                Vec::new()
+            },
+            rows: if num_rows > 0 {
+                Vec::with_capacity(num_rows)
+            } else {
+                Vec::new()
+            },
+            column_indexes: if num_columns > 0 {
+                HashMap::with_capacity(num_columns)
+            } else {
+                HashMap::new()
+            },
```

Comment:
You can simplify this to `with_capacity(num_x)` unconditionally -- it's equal to `new()` when capacity is 0 https://doc.rust-lang.org/src/alloc/raw_vec.rs.html#120-123